examples/waveShared.cpp

```diff
@@ -0,0 +1,441 @@
+    RAJA::NestedPolicy<RAJA::ExecList<RAJA::seq_exec, RAJA::seq_exec>>;  
+#else
+
+  RAJA::RangeSegment outerRange(0, grid.nx);
+
+  using abstractPolicy
+    = RAJA::NestedPolicy<RAJA::ExecList
+    <RAJA::cuda_threadblock_y_exec<By>,
+     RAJA::cuda_threadblock_x_exec<Bx>>>;
+#endif
+
+  time = 0;
+  setIC(P1, P2, (time - dt), time, grid);
+  for (int k = 0; k < nt; ++k) {
+    waveShared<double, abstractPolicy>(P1, P2, outerRange, ct, grid.nx);
+
+    time += dt;
+    double *Temp = P2;
+
+    P2 = P1;
+    P1 = Temp;
+  }
+
+#if defined(RAJA_ENABLE_CUDA)
+  cudaDeviceSynchronize();
+#endif
+  computeErr(P2, time, grid);
+  printf("Evolved solution to time = %f \n", time);
+
+
+  memoryManager::deallocate(P1);
+  memoryManager::deallocate(P2);
+
+  return 0;
+}
+
+
+/*
+  Function for the analytic solution
+*/
+double waveSol(double t, double x, double y)
+{
+  return cos(2. * PI * t) * sin(2. * PI * x) * sin(2. * PI * y);
+}
+
+/*
+  Error is computed via ||P_{approx}(:) - P_{analytic}(:)||_{inf}
+*/
+void computeErr(double *P, double tf, grid_s grid)
+{
+
+  RAJA::RangeSegment fdBounds(0, grid.nx);
+  RAJA::ReduceMax<RAJA::seq_reduce, double> tMax(-1.0);
+  using myPolicy =
+    RAJA::NestedPolicy<RAJA::ExecList<RAJA::seq_exec, RAJA::seq_exec>>;
+
+  RAJA::forallN<myPolicy>(
+    fdBounds, fdBounds, [=](RAJA::Index_type ty, RAJA::Index_type tx) {
+
+      int id = tx + grid.nx * ty;
+      double x = grid.ox + tx * grid.dx;
+      double y = grid.ox + ty * grid.dx;
+      double myErr = std::abs(P[id] - waveSol(tf, x, y));
+
+      /*
+        tMax.max() is used to store the maximum value
+      */
+      tMax.max(myErr);
+    });
+
+  double lInfErr = tMax;
+  printf("Max Error = %lg, dx = %f \n", lInfErr, grid.dx);
+}
+
+
+/*
+ Function to set intial condition
+*/
+void setIC(double *P1, double *P2, double t0, double t1, grid_s grid)
+{
+
+  using myPolicy =
+    RAJA::NestedPolicy<RAJA::ExecList<RAJA::seq_exec, RAJA::seq_exec>>;
+  RAJA::RangeSegment fdBounds(0, grid.nx);
+  
+  RAJA::forallN<myPolicy>(
+    fdBounds, fdBounds, [=](RAJA::Index_type ty, RAJA::Index_type tx) {    
+
+      int id = tx + ty * grid.nx;
+      double x = grid.ox + tx * grid.dx;
+      double y = grid.ox + ty * grid.dx;
+      
+      P1[id] = waveSol(t0, x, y);
+      P2[id] = waveSol(t1, x, y);
+    });
+}
+
+
+//with shared memory
+template <typename T, typename fdNestedPolicy>
+void waveShared(T *P1, T *P2, RAJA::RangeSegment outerBounds, double ct, int nx)
+{
+
+#if CPU
+  //RAJA will give outer Id 
+ RAJA::forallN<fdNestedPolicy>(
+      outerBounds, outerBounds, [=] (RAJA::Index_type outerIdy, RAJA::Index_type outerIdx) {
+#else
+  //RAJA will give global id
+  RAJA::forallN<fdNestedPolicy>(
+      outerBounds, outerBounds, [=] __device__ (RAJA::Index_type gIdy, RAJA::Index_type gIdx) {
+        int outerIdx = blockIdx.x; //What I really want is the block id instead of thread of id... 
+        int outerIdy = blockIdx.y; //stick these in here...
```

Comment:
@davidbeckingsale, here I would like to modify the indices to traverse though cuda block Id's instead of the the thread id's. These loops could be referred to as the outer loops.